pkg/storaged/btrfs/btrfs-tool.py

```diff
@@ -0,0 +1,255 @@
+#! /usr/bin/python3
+
+# btrfs-tool  --  Query and monitor btrfs filesystems
+#
+# This program monitors all btrfs filesystems and reports their
+# subvolumes and other things.
+#
+# It can do that continously, or as a one shot operation.  The tool
+# mounts btrfs filesystems as necessary to retrieve the requested
+# information, but does it in a polite way: they are mounted once and
+# then left mounted until that is no longer needed. Typically, you
+# might see some mounts when a Cockpit session starts, and the
+# corresponding unmounts when it ends.
+#
+# This tool can be run multiple times concurrently with itself, and it
+# wont get confused.
+
+import contextlib
+import fcntl
+import json
+import os
+import re
+import signal
+import subprocess
+import sys
+import time
+
+
+def debug(msg):
+    # subprocess.check_call(["logger", msg])
+    pass
+
+
+TMP_MP_DIR = "/var/lib/cockpit/btrfs"
+
+
+@contextlib.contextmanager
+def mount_database():
+    path = TMP_MP_DIR + "/db"
+    os.makedirs(TMP_MP_DIR, mode=0o700, exist_ok=True)
+    fd = os.open(path, os.O_RDWR | os.O_CREAT)
+    fcntl.flock(fd, fcntl.LOCK_EX)
+    data = os.read(fd, 100000)
+    blob = json.loads(data) if len(data) > 0 else {}
+    try:
+        yield blob
+        data = json.dumps(blob).encode() + b"\n"
+        os.lseek(fd, 0, os.SEEK_SET)
+        os.truncate(fd, 0)
+        os.write(fd, data)
+    finally:
+        os.close(fd)
+
+
+@contextlib.contextmanager
+def context_chdir(path):
+    old_cwd = os.getcwd()
+    os.chdir(path)
+    try:
+        yield
+    finally:
+        os.chdir(old_cwd)
+
+
+def list_filesystems():
+    output = json.loads(subprocess.check_output(["lsblk", "-Jplno", "NAME,FSTYPE,UUID,MOUNTPOINTS"]))
+    filesystems = {}
+    for b in output['blockdevices']:
+        if b['fstype'] == "btrfs":
+            uuid = b['uuid']
+            mps = list(filter(lambda x: x is not None and not x.startswith(TMP_MP_DIR), b['mountpoints']))
+            has_tmp_mp = len(list(filter(lambda x: x is not None and x.startswith(TMP_MP_DIR), b['mountpoints']))) > 0
+            if uuid not in filesystems:
+                filesystems[uuid] = {
+                    'uuid': uuid,
+                    'devices': [b['name']],
+                    'mountpoints': mps,
+                    'has_tmp_mountpoint': has_tmp_mp
+                }
+            else:
+                filesystems[uuid]['devices'] += [b['name']]
+                filesystems[uuid]['mountpoints'] += mps
+                filesystems[uuid]['has_tmp_mountpoint'] = filesystems[uuid]['has_tmp_mountpoint'] or has_tmp_mp
+    return filesystems
+
+
+tmp_mountpoints = set()
```

Comment:
Not looking too deep, we have this global state and the `db` why do we not save the tmp_mountpoints in the DB?

Reply:
tmp_mountpoints is per process state, just so that we remember which ones this very process has added, so that they can all be removed when it exits. The db is global state for all processes.